File: packages/sdk-node/src/openai.ts
import type { Regressa } from "./client.js";
import type { CallOptions, Message } from "./types.js";
import { contentToText, inferTemplate } from "./template.js";

/**
 * Structural type for the OpenAI client — we don't import `openai` so the SDK stays dependency-free.
 */
interface OpenAILike {
  chat: { completions: { create: (...args: any[]) => any } };
}

type CreateParams = { model: string; messages: Message[]; stream?: boolean; regressa?: CallOptions; [k: string]: unknown };

/**
 * Wrap an OpenAI client. Returns the same client with `chat.completions.create` instrumented.
 *
 *   const openai = wrapOpenAI(new OpenAI(), regressa);
 *   await openai.chat.completions.create({ model, messages, regressa: { promptTemplate: { name, raw } } });
 */
export function wrapOpenAI<T extends OpenAILike>(client: T, regressa: Regressa, defaults: CallOptions = {}): T {
  const original = client.chat.completions.create.bind(client.chat.completions);

  client.chat.completions.create = (params: CreateParams, ...rest: unknown[]) => {
    const { regressa: callOpts, ...clean } = params;
    const opts: CallOptions = { ...defaults, ...callOpts };
    const started = Date.now();
    const base = {
      model: clean.model,
      provider: "openai" as const,
      input_messages: clean.messages,
      prompt_template: opts.promptTemplate ?? inferTemplate(clean.messages),
      trace_group_id: opts.traceGroupId,
      metadata: opts.metadata,
    };

    const promise = original(clean, ...rest);
    if (clean.stream) return wrapStream(promise, regressa, base, started);

    return promise.then(
      (res: any) => {
        regressa.trace({
          ...base,
          model: res?.model ?? clean.model,
          output_text: res?.choices?.[0]?.message?.content ?? null,
          prompt_tokens: res?.usage?.prompt_tokens,
          completion_tokens: res?.usage?.completion_tokens,
          latency_ms: Date.now() - started,
          status: "success",
        });
        return res;
      },
      (err: unknown) => {
        regressa.trace({ ...base, latency_ms: Date.now() - started, status: "error", error_message: errMsg(err) });
        throw err;
      },
    );
  };
  return client;
}

async function wrapStream(promise: Promise<any>, regressa: Regressa, base: any, started: number) {
  const stream = await promise;
  const chunks: string[] = [];
  let usage: any = null;
  let model: string | undefined;
  const iter = stream[Symbol.asyncIterator].bind(stream);
  stream[Symbol.asyncIterator] = async function* () {
    try {
      for await (const chunk of iter()) {
        model = chunk?.model ?? model;
        const delta = chunk?.choices?.[0]?.delta?.content;
        if (delta) chunks.push(delta);
        if (chunk?.usage) usage = chunk.usage;
        yield chunk;
      }
      regressa.trace({ ...base, model: model ?? base.model, output_text: chunks.join(""),
        prompt_tokens: usage?.prompt_tokens, completion_tokens: usage?.completion_tokens,
        latency_ms: Date.now() - started, status: "success" });
    } catch (err) {
      regressa.trace({ ...base, output_text: chunks.join(""), latency_ms: Date.now() - started, status: "error", error_message: errMsg(err) });
      throw err;
    }
  };
  return stream;
}

function errMsg(err: unknown) { return err instanceof Error ? err.message : String(err); }
export { contentToText };

File: packages/sdk-python/regressa/openai.py
from __future__ import annotations

import time
from typing import Any

from .client import Regressa, is_timeout
from .template import infer_template


def wrap_openai(client: Any, regressa: Regressa, **defaults: Any) -> Any:
    """Instrument an `openai.OpenAI` client's chat.completions.create (sync, incl. stream=True).

    Extra per-call kwargs understood by Regressa (stripped before hitting OpenAI):
        regressa_prompt_template={"name": ..., "raw": ...}, regressa_trace_group_id=..., regressa_metadata={...}
    """
    original = client.chat.completions.create

    def create(*args: Any, **kwargs: Any) -> Any:
        opts = _pop_opts(kwargs, defaults)
        messages = kwargs.get("messages", [])
        base: dict[str, Any] = {
            "model": kwargs.get("model", "unknown"),
            "provider": "openai",
            "input_messages": messages,
            "prompt_template": opts.get("prompt_template") or infer_template(messages),
            "trace_group_id": opts.get("trace_group_id"),
            "metadata": opts.get("metadata"),
        }
        started = time.perf_counter()
        try:
            res = original(*args, **kwargs)
        except Exception as exc:
            regressa.trace(
                **base,
                latency_ms=_ms(started),
                status="timeout" if is_timeout(exc) else "error",
                error_message=str(exc),
            )
            raise
        if kwargs.get("stream"):
            return _StreamProxy(res, regressa, base, started, _openai_chunk)
        usage = getattr(res, "usage", None)
        choice = res.choices[0] if getattr(res, "choices", None) else None
        regressa.trace(
            **{**base, "model": getattr(res, "model", None) or base["model"]},
            output_text=getattr(getattr(choice, "message", None), "content", None),
            prompt_tokens=getattr(usage, "prompt_tokens", None),
            completion_tokens=getattr(usage, "completion_tokens", None),
            latency_ms=_ms(started),
            status="success",
        )
        return res

    client.chat.completions.create = create
    return client


def _openai_chunk(chunk: Any, state: dict[str, Any]) -> None:
    if getattr(chunk, "model", None):
        state["model"] = chunk.model
    choices = getattr(chunk, "choices", None) or []
    if choices and getattr(choices[0], "delta", None) is not None:
        delta = getattr(choices[0].delta, "content", None)
        if delta:
            state["chunks"].append(delta)
    usage = getattr(chunk, "usage", None)
    if usage is not None:
        state["prompt_tokens"] = getattr(usage, "prompt_tokens", None)
        state["completion_tokens"] = getattr(usage, "completion_tokens", None)


class _StreamProxy:
    """Iterates the underlying stream, records a trace when it finishes (or errors)."""

    def __init__(self, stream: Any, regressa: Regressa, base: dict[str, Any], started: float, on_chunk: Any) -> None:
        self._stream, self._regressa, self._base, self._started, self._on_chunk = (
            stream,
            regressa,
            base,
            started,
            on_chunk,
        )
        self._state: dict[str, Any] = {"chunks": [], "model": None, "prompt_tokens": None, "completion_tokens": None}

    def __iter__(self):
        try:
            for chunk in self._stream:
                self._on_chunk(chunk, self._state)
                yield chunk
        except Exception as exc:
            self._regressa.trace(
                **self._base,
                output_text="".join(self._state["chunks"]),
                latency_ms=_ms(self._started),
                status="error",
                error_message=str(exc),
            )
            raise
        s = self._state
        self._regressa.trace(
            **{**self._base, "model": s["model"] or self._base["model"]},
            output_text="".join(s["chunks"]),
            prompt_tokens=s["prompt_tokens"],
            completion_tokens=s["completion_tokens"],
            latency_ms=_ms(self._started),
            status="success",
        )

    def __getattr__(self, item: str) -> Any:
        return getattr(self._stream, item)


def _pop_opts(kwargs: dict[str, Any], defaults: dict[str, Any]) -> dict[str, Any]:
    opts = dict(defaults)
    for key in ("prompt_template", "trace_group_id", "metadata"):
        val = kwargs.pop(f"regressa_{key}", None)
        if val is not None:
            opts[key] = val
    return opts


def _ms(started: float) -> int:
    return int((time.perf_counter() - started) * 1000)
